vending-app/src/main/java/org/microg/vending/ui/VendingActivity.kt

```diff
@@ -0,0 +1,324 @@
+                            )
+                        }
+                    }.onFailure { Log.i(TAG, "couldn't purchase ${app.packageName}: ${it.message}") }
+                        .onSuccess { Log.d(TAG, "purchased ${app.packageName} successfully") }
+
+                    // Get download links for requested package
+                    val downloadUrls = runCatching {
+
+                        client.requestDownloadUrls(
+                        app.packageName,
+                        app.versionCode!!.toLong(),
+                        auth!!,
+                        deliveryToken = app.deliveryToken
+                    ) }
+
+                    if (downloadUrls.isFailure) {
+                        Log.w(TAG, "Failed to request download URLs: ${downloadUrls.exceptionOrNull()!!.message}")
+                        apps[app] = previousState
+                        return@runBlocking
+                    }
+
+                    runCatching {
+
+                        var lastNotification = 0L
+                        installPackagesFromNetwork(
+                            packageName = app.packageName,
+                            components = downloadUrls.getOrThrow(),
+                            httpClient = client,
+                            isUpdate = isUpdate
+                        ) { session, progress ->
+
+
+                            // Android rate limits notification updates by some vague rule of "not too many in less than one second"
+                            if (progress !is Downloading || lastNotification + 250 < System.currentTimeMillis()) {
+                                notifyInstallProgress(app.displayName, session, progress)
+                                lastNotification = System.currentTimeMillis()
+                            }
+
+                            if (progress is Downloading) apps[app] = progress
+                            else if (progress is CommitingSession) apps[app] = Pending
+                        }
+                    }.onSuccess {
+                        apps[app] = Installed
+                    }.onFailure { exception ->
+                        Log.w(TAG, "Installation from network unsuccessful.", exception)
+                        apps[app] = previousState
+                    }
+                }
+            }.start()
+        }
+
+        val uninstall: (app: EnterpriseApp) -> Unit = { app ->
+            Thread {
+                runBlocking {
+
+                    val previousState = apps[app]!!
+                    apps[app] = Pending
+                    runCatching { uninstallPackage(app.packageName) }.onSuccess {
+                        apps[app] = NotInstalled
+                    }.onFailure {
+                        apps[app] = previousState
+                    }
+
+                }
+            }.start()
+        }
+
+        setContent {
+            VendingUi(account, install, uninstall)
+        }
```

Comment:
There is a lot of logic in the onCreate() callback. Maybe we can extract some part into separate methods ?